include/cantera/numerics/AdaptivePreconditioner.h

```diff
@@ -0,0 +1,300 @@
+/**
+ *  @file AdaptivePreconditioner.h
+ *   Declarations for the class AdaptivePreconditioner
+ *   which is a child class of PreconditionerBase for preconditioners used by sundials
+ */
+
+// This file is part of Cantera. See License.txt in the top-level directory or
+// at https://cantera.org/license.txt for license and copyright information.
+
+#ifndef ADAPTIVEPRECONDITIONER_H
+#define ADAPTIVEPRECONDITIONER_H
+
+#if CT_USE_SYSTEM_EIGEN
+#include <Eigen/Sparse>
+#else
+#include "cantera/ext/Eigen/Sparse"
+#endif
+
+#include "cantera/numerics/PreconditionerBase.h"
+#include "cantera/kinetics/ReactionDerivativeManager.h"
+#include "cantera/zerodim.h"
+#include "cantera/thermo.h"
+#include "cantera/kinetics.h"
+#include "cantera/kinetics/Reaction.h"
+
+#include "float.h"
+#include <unordered_map>
+
+namespace Cantera
+{
+
+//! Flag to indicate adaptive preconditioner is set
+const int ADAPTIVE_MECHANISM_PRECON_MATRIX = 1;
+
+class AdaptivePreconditioner : public PreconditionerBase
+{
+    protected:
+        //! @param m_values is the container for the values that are mapped to m_jacobian
+        std::vector<double> m_values;
+
+        //! @param m_inner is the container for inner (row) indices in CSC format
+        std::vector<int> m_inner;
+
+        //! @param m_outer is the container for outer starts (nnz elements per row)
+        std::vector<int> m_outer;
+
+        //! @param m_sizes is the container for starts of each reactor in double
+        std::vector<size_t> m_sizes;
+
+        //! @param m_reaction_derv_mgrs vector of reaction derivative managers for each kinetic object
+        std::vector<ReactionDerivativeManager> m_reaction_derv_mgrs;
+
+        //! @param m_thermo_indices indices of thermo variables temp, mass, vol, etc.
+        std::vector<size_t> m_thermo_indices;
+
+        //! @param m_state_indices total index map of system, the key is a flattened index and the output is the corresponding derivative index
+        std::unordered_map<int, int> m_state_indices;
+
+        //! @param m_nnz is the number of non zeros
+        int m_nnz = 0;
```

Comment:
I don't think `@param` should be used for documenting member variables. I would generally suggest just going with something like
```suggestion
        //! The number of non zeros
        int m_nnz = 0;
```